R/messages.R

```diff
@@ -110,3 +110,64 @@ setvar <- function(core, dates, var, values, unit)
 
     invisible(NULL)
 }
+
+
+#' Fetch all the available Hector variables
+#'
+#' This function is similar to fetchvars, except it fetches all available
+#' Hector variables, including exogenous inputs (e.g., emissions & pre-industrial
+#' concentrations), parameters (e.g., alpha, beta, ECS), and output
+#' variables (e.g., concentrations & forcings).
+#'
+#' @param core Hector core object
+#' @param dates Vector of dates, optional
+#' @param scenario Scenario name, optional str
+#' @param outfile Absolute path, including the name, of the output .csv file.
+#' @return vars_all    Dataframe containing all Hector variables
+#' @family main user interface functions
+#' @export
+fetchvars_all <- function(core, dates=NULL, scenario=NULL, outfile=NULL)
+{
+
+    # Get output for variables that do use the date param
+    rslt_date <- fetchvars(core, dates, vars_date, scenario)
+
+    # Get variables that DO NOT use date arg
+    rslt_nodate <- fetchvars(core, NA, vars_nodate, scenario)
+
+    rslt <- rbind(rslt_date, rslt_nodate)
+
+    if (!is.null(outfile)) {
+        # Check that outfile ends with the .csv file extension
+        if (substr(outfile, nchar(outfile)-3, nchar(outfile)) != ".csv") {
+            stop("'outfile' must end in '.csv'")
+        }
+
+        # Pivot the dataframe "wide" so years run along the x axis
+        # Catch warning from reshape()
+        suppressWarnings(rslt_pretty <- reshape(rslt, direction="wide",
+                                        idvar=c("scenario", "variable", "units"),
+                                        timevar="year"))
+
+        # Clean up the column names (value.YYYY --> YYYY)
+        col_names <- colnames(rslt_pretty)[-1:-3]
+        col_names <- sapply(col_names, function(x) sub("value.", "", x), USE.NAMES=FALSE)
+        col_names <- col_names[-length(col_names)]
+
+        col_names <- c("scenario", "variable", "units", "initial_value", col_names)
+
+        num_cols <- length(col_names)
+
+        # Move last column (initial_value) into 4th col position & scenario into 1st col
+        rslt_pretty <- rslt_pretty[, c(1, 2, 3, num_cols, 5:num_cols-1)]
```

Comment:
Using `5:num_cols-1` is error-prone and confusing to anyone reading the code.  It's hard to know just by looking at it whether you really meant 4 ... nc-1, or whether you meant 5 ... nc-1 and made an error.  It's much clearer to write `seq(4, nc-1)` because that leaves no doubt.